File: sdk-java/src/main/java/io/superdurable/dex/SubFlow.java
/*
 * Copyright (c) 2026 Super Durable, Inc.
 *
 * Licensed under the Super Durable Source License 1.0.
 * You may not use this file except in compliance with the License.
 * See the LICENSE file in the repository root.
 *
 * SPDX-License-Identifier: LicenseRef-Super-Durable-1.0
 */

package io.superdurable.dex;

/**
 * Creates durable SubFlow conditions and reads their results during Step execution.
 *
 * <p>A SubFlow is a normal, independently addressable Flow. Returning a condition created by
 * {@link #run} from {@link Step#waitFor} starts or reuses that Flow and makes its completion part
 * of the surrounding {@link Wait}. Read the corresponding result only from the same Step's
 * {@link Step#execute} method.
 *
 * <pre>{@code
 * public Wait waitFor(Context context, Order input) {
 *     return Wait.until(SubFlow.run(ChargeOrderFlow.class, input));
 * }
 *
 * public StepDecision execute(Context context, Order input) {
 *     FlowResult charge = SubFlow.getConditionResults(context);
 *     return StepDecision.gracefulComplete(charge.getSingleOutput(Receipt.class));
 * }
 * }</pre>
 */
public final class SubFlow {
    private SubFlow() {
    }

    /**
     * Creates a SubFlow condition using default options.
     *
     * @param flowClass the registered SubFlow class
     * @param input the typed input for its starting Step
     * @param <I> the starting input type
     * @param <F> the Flow implementation type
     * @return a durable SubFlow condition
     * @throws IllegalArgumentException if {@code flowClass} is {@code null}
     */
    @SuppressWarnings("unchecked")
    public static <I, F extends Flow<I>> Condition run(
            final Class<F> flowClass,
            final I input) {
        return Condition.subFlow((Class<? extends Flow<?>>) (Class<?>) flowClass, input, null);
    }

    /**
     * Creates a SubFlow condition with explicit options.
     *
     * @param flowClass the registered SubFlow class
     * @param input the typed input for its starting Step
     * @param options start, reuse, and condition options
     * @param <I> the starting input type
     * @param <F> the Flow implementation type
     * @return a durable SubFlow condition
     * @throws IllegalArgumentException if {@code flowClass} is {@code null}
     */
    @SuppressWarnings("unchecked")
    public static <I, F extends Flow<I>> Condition run(
            final Class<F> flowClass,
            final I input,
            final SubFlowOptions options) {
        return Condition.subFlow(
                (Class<? extends Flow<?>>) (Class<?>) flowClass, input, options);
    }

    /**
     * Returns the first SubFlow condition result from this Step invocation.
     *
     * @param context the current execute invocation context
     * @return the running snapshot or terminal result at SubFlow index zero
     * @throws IllegalStateException if called outside the Step's execute invocation
     * @throws IllegalArgumentException if the context is not managed by Dex or no SubFlow exists
     */
    public static FlowResult getConditionResults(final Context context) {
        return getConditionResults(context, 0);
    }

    /**
     * Returns one SubFlow condition result from this Step invocation.
     *
     * @param context the current execute invocation context
     * @param index the zero-based SubFlow condition index
     * @return the running snapshot or terminal result
     * @throws IllegalStateException if called outside the Step's execute invocation
     * @throws IllegalArgumentException if the context is not managed by Dex or the index is invalid
     */
    public static FlowResult getConditionResults(final Context context, final int index) {
        if (!(context instanceof InvocationContext)) {
            throw new IllegalArgumentException("SubFlow results require a Dex invocation Context");
        }
        return ((InvocationContext) context).subFlowResult(index);
    }

    /**
     * Returns the generated Flow ID for the first SubFlow condition in this Step invocation.
     *
     * <p>The ID starts with {@code SubFlow:} and is available for running and terminal results.
     * It can be passed to
     * {@link Client#stopFlow(String)} when another condition wins a {@link Wait#anyOf}.
     *
     * @param context the current execute invocation context
     * @return the generated SubFlow Flow ID at index zero
     * @throws IllegalStateException if called outside the Step's execute invocation
     * @throws IllegalArgumentException if the context is not managed by Dex or no SubFlow exists
     */
    public static String getFlowId(final Context context) {
        return getFlowId(context, 0);
    }

    /**
     * Returns the generated Flow ID for one SubFlow condition in this Step invocation.
     *
     * <p>The zero-based index uses the stable order of SubFlow conditions in the surrounding
     * {@link Wait}. The ID remains valid after the parent Step completes.
     *
     * @param context the current execute invocation context
     * @param index the zero-based SubFlow condition index
     * @return the generated SubFlow Flow ID
     * @throws IllegalStateException if called outside the Step's execute invocation
     * @throws IllegalArgumentException if the context is not managed by Dex or the index is invalid
     */
    public static String getFlowId(final Context context, final int index) {
        if (!(context instanceof InvocationContext)) {
            throw new IllegalArgumentException("SubFlow IDs require a Dex invocation Context");
        }
        return ((InvocationContext) context).subFlowId(index);
    }
}

File: sdk-rust/crates/dex-sdk/src/sub_flow.rs
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

use std::any::TypeId;
use std::time::Duration;

use crate::start_flow_options::InitialAttribute;
use crate::wait::SubFlowDefinition;
use crate::{
    Attribute, AttributeMap, Condition, Context, Flow, FlowConfig, FlowResult, FlowTimeoutPolicy,
    HandlerResult, RetryPolicy, SdkResult, Value, value_mapper,
};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
/// Controls how a generated SubFlow Flow ID resolves an existing execution.
pub enum SubFlowReusePolicy {
    /// Attaches to a running execution or returns its existing terminal result.
    Attach,
    /// Restarts abnormal executions, attaches while running, and returns completed results.
    RestartIfPreviousExitsAbnormally,
    /// Replaces any different existing execution, including a running one.
    AlwaysRestart,
}

#[derive(Clone)]
/// Configures one durable SubFlow Condition.
///
/// The SubFlow inherits its parent's effective Flow configuration. Dex generates its Flow ID and
/// request ID. Builder methods preserve normal Flow start defaults when omitted.
pub struct SubFlowOptions {
    pub(crate) timeout: Option<Duration>,
    pub(crate) timeout_policy: FlowTimeoutPolicy,
    pub(crate) start_delay: Option<Duration>,
    pub(crate) retry_policy: Option<RetryPolicy>,
    pub(crate) config_override: Option<FlowConfig>,
    pub(crate) attributes: Vec<InitialAttribute>,
    pub(crate) reuse_policy: SubFlowReusePolicy,
    pub(crate) condition_id: Option<String>,
}

impl SubFlowOptions {
    /// Creates options using the abnormal-restart reuse policy and normal start defaults.
    pub fn new() -> Self {
        Self {
            timeout: None,
            timeout_policy: FlowTimeoutPolicy::Default,
            start_delay: None,
            retry_policy: None,
            config_override: None,
            attributes: Vec::new(),
            reuse_policy: SubFlowReusePolicy::RestartIfPreviousExitsAbnormally,
            condition_id: None,
        }
    }

    /// Sets the maximum total SubFlow execution duration.
    pub fn timeout(mut self, value: Duration) -> Self {
        self.timeout = Some(value);
        self
    }

    /// Selects what Dex does when the positive soft SubFlow timeout expires.
    pub fn timeout_policy(mut self, value: FlowTimeoutPolicy) -> Self {
        self.timeout_policy = value;
        self
    }

    /// Delays the SubFlow starting Step after server acceptance.
    pub fn start_delay(mut self, value: Duration) -> Self {
        self.start_delay = Some(value);
        self
    }

    /// Sets whole-Flow retry behavior after abnormal completion.
    pub fn retry_policy(mut self, value: RetryPolicy) -> Self {
        self.retry_policy = Some(value);
        self
    }

    /// Applies fields over the inherited parent Flow configuration.
    pub fn config_override(mut self, value: FlowConfig) -> Self {
        self.config_override = Some(value);
        self
    }

    /// Selects existing-execution resolution behavior.
    pub fn reuse_policy(mut self, value: SubFlowReusePolicy) -> Self {
        self.reuse_policy = value;
        self
    }

    /// Assigns the stable ID required by [`crate::Wait::any_combination_of`].
    pub fn condition_id(mut self, value: impl Into<String>) -> Self {
        self.condition_id = Some(value.into());
        self
    }

    /// Adds one initial singleton Attribute owned by the target SubFlow.
    pub fn initial_attribute<T: Value>(mut self, attribute: &Attribute<T>, value: T) -> Self {
        self.attributes
            .push(InitialAttribute::for_attribute(attribute, value));
        self
    }

    /// Adds one initial Attribute-map value for an instance.
    /// Slash is prohibited in instance keys because it is a reserved character.
    pub fn initial_attribute_map<T: Value>(
        mut self,
        attribute: &AttributeMap<T>,
        instance: &str,
        value: T,
    ) -> Self {
        self.attributes.push(InitialAttribute::for_attribute_map(
            attribute, instance, value,
        ));
        self
    }
}

impl Default for SubFlowOptions {
    fn default() -> Self {
        Self::new()
    }
}

/// Creates durable SubFlow Conditions and reads their Execute results.
pub struct SubFlow;

impl SubFlow {
    /// Creates a SubFlow Condition with default options.
    ///
    /// # Errors
    ///
    /// Returns a value-mapping error when `input` cannot be encoded.
    pub fn run<SomeFlow: Flow>(
        flow: &SomeFlow,
        input: SomeFlow::StartInput,
    ) -> SdkResult<Condition> {
        Self::run_with_options(flow, input, SubFlowOptions::new())
    }

    /// Creates a SubFlow Condition with explicit options.
    ///
    /// # Errors
    ///
    /// Returns a value-mapping error when `input` cannot be encoded. Worker mapping later validates
    /// the exact registered Rust Flow type, starting Step, Attributes, and option durations.
    pub fn run_with_options<SomeFlow: Flow>(
        flow: &SomeFlow,
        input: SomeFlow::StartInput,
        options: SubFlowOptions,
    ) -> SdkResult<Condition> {
        Ok(Condition::sub_flow(SubFlowDefinition {
            flow_type: flow.flow_type(),
            type_id: TypeId::of::<SomeFlow>(),
            input: value_mapper::encode(&input)?,
            options,
        }))
    }

    /// Returns the first SubFlow result during Step `execute`.
    pub fn condition_result(context: &Context) -> HandlerResult<FlowResult> {
        Self::condition_result_at(context, 0)
    }

    /// Returns one stable-indexed SubFlow result during Step `execute`.
    pub fn condition_result_at(context: &Context, index: usize) -> HandlerResult<FlowResult> {
        context.sub_flow_result(index)
    }

    /// Returns the generated Flow ID for the first SubFlow Condition during Step `execute`.
    pub fn flow_id(context: &Context) -> HandlerResult<String> {
        Self::flow_id_at(context, 0)
    }

    /// Returns the generated Flow ID for one stable-indexed SubFlow Condition during `execute`.
    pub fn flow_id_at(context: &Context, index: usize) -> HandlerResult<String> {
        context.sub_flow_id(index)
    }
}
